include/boost/url/impl/any_allocator.hpp

```diff
@@ -0,0 +1,183 @@
+}
+
+template <std::size_t N, class Allocator>
+typename std::enable_if<N < alignof(max_align_t), void>::type
+try_deallocate_aligned(const Allocator& a, char* p, std::size_t n, std::size_t size, std::size_t align) {
+    if (align <= N) {
+        using aligned_t = typename boost::type_with_alignment<N>::type;
+        using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+        aligned_allocator_t(a).deallocate(reinterpret_cast<aligned_t*>(p), n * ceil_div(size, align));
+    } else {
+        try_deallocate_aligned<2*N>(a, p, n, size, align);
+    }
+}
+
+template <class Allocator>
+void deallocate_aligned(const Allocator& a, char* p, std::size_t n, std::size_t size, std::size_t align) {
+    try_deallocate_aligned<1>(a, p, n, size, align);
+}
+}
+
+template <class T>
+template <class Allocator>
+any_allocator<T>::
+any_allocator(Allocator const& a) noexcept
+{
+    class impl;
+
+    using char_allocator_type = boost::allocator_rebind_t<Allocator, char>;
+
+    class impl : public detail::any_allocator_base
+    {
+         char_allocator_type a_;
+
+    public:
+        explicit
+        impl(char_allocator_type const& a)
+        : any_allocator_base(typeid(char_allocator_type)), a_(a)
+        {
+        }
+
+        char*
+        allocate(std::size_t n, std::size_t size, std::size_t align) override
+        {
+            if (size == 1) {
+                return a_.allocate(n);
+            } else {
+                return detail::allocate_aligned(a_, n, size, align);
+            }
+        }
+
+        void
+        deallocate(char* p, std::size_t n, std::size_t size, std::size_t align)
+        {
+            if (size == 1) {
+                a_.deallocate(p, n);
+            } else {
+                detail::deallocate_aligned(a_, p, n, size, align);
+            }
+        }
+
+        bool
+        is_equal(const any_allocator_base& other) override {
+            const impl* other_impl = static_cast<const impl*>(&other);
+            if (!other_impl) {
+                return false;
+            }
+            return a_ == other_impl->a_;
+        };
+    };
+
+    p_ = std::allocate_shared<impl>(a, char_allocator_type{});
+}
+
+template <class T>
+template <class U>
+any_allocator<T>::
+any_allocator(
+    any_allocator<U> const& other) noexcept
+: p_(other.p_)
+{
+}
+
+template <class T>
+template <class U>
+any_allocator<T>::
+any_allocator(
+    any_allocator<U> && other) noexcept
+: p_(std::move(other.p_))
+{
+}
+
+template <class T>
+T*
+any_allocator<T>::
+allocate(
+    std::size_t n) const
+{
+    if (p_ != nullptr)
+        return reinterpret_cast<T*>(p_->allocate(n, sizeof(T), alignof(T)));
+    return std::allocator<T>().allocate(n);
+}
+
+template <class T>
+void
+any_allocator<T>::
+deallocate(
+    T* p,
+    std::size_t n)
+{
+    if (p_ != nullptr)
+        return p_->deallocate(reinterpret_cast<char*>(p), n, sizeof(T), alignof(T));
```

Comment:
The internal allocator is bound to `char` so I just kept the convention. Shouldn't it work either way?

Reply:
it should, but then you wont have to reinterpret_cast

Reply:
I do need to reinterpret_cast or static_cast everywhere anyway but you're right. `void*` does have better semantics in the API because it makes it clear the type is erased.